widechar_width.py

```diff
@@ -0,0 +1,1507 @@
+# widechar_width.py for Unicode 14.0.0
+# See https://github.com/ridiculousfish/widecharwidth/
+#
+# SHA1 file hashes:
+#  (
+#  the hashes for generate.py and the template are git object hashes,
+#  use `git log --all --find-object=<hash>` in the widecharwidth repository
+#  to see which commit they correspond to,
+#  or run `git hash-object` on the file to compare.
+#  The other hashes are simple `sha1sum` style hashes.
+#  )
+#
+#  generate.py:         4dc82efc75ba2c3e77ffb9832dd78559f3c02e10
+#  template.js:         e41a73aa3593bba724d87ad7eb3c66635ba17e6c
+#  UnicodeData.txt:     8a5c26bfb27df8cfab23cf2c34c62d8d3075ae4d
+#  EastAsianWidth.txt:  8ec36ccac3852bf0c2f02e37c6151551cd14db72
+#  emoji-data.txt:      3f0ec08c001c4bc6df0b07d01068fc73808bfb4c
+
+from typing import Union
+from enum import IntEnum
+
+# Special width values
+class Special(IntEnum):
+    nonprint = -1  # The character is not printable.
+    combining = -2  # The character is a zero-width combiner.
+    ambiguous = -3  # The character is East-Asian ambiguous width.
+    private_use = -4  # The character is for private use.
+    unassigned = -5  # The character is unassigned.
+    widened_in_9 = -6  # Width is 1 in Unicode 8, 2 in Unicode 9+.
+    non_character = -7  # The character is a noncharacter.
+
+
+class Codepointrange:
+    def __init__(self, *ranges):
+        self.ranges = sorted(ranges)
+
+    def __contains__(self, char):
+        left = 0
+        right = len(self.ranges) - 1
+        if char < self.ranges[0][0] or char > self.ranges[right][1]:
+            return False
+
+        while right >= left:
+            middle = (left + right) // 2
+
+            if char < self.ranges[middle][0]:
+                right = middle - 1
+            elif char > self.ranges[middle][1]:
+                left = middle + 1
+            else:
+                return True
+        return False
+
+
+# Simple ASCII characters - used a lot, so we check them first.
+ascii_table = Codepointrange(
```

Comment:
the Rust thing has global constants uppercase, not sure if that convention also exists for python (where there are no real constants AFAIK)

Reply:
[PEP8](https://www.python.org/dev/peps/pep-0008/#naming-conventions) indeed says:

> Constants are usually defined on a module level and written in all capital letters with underscores separating words. Examples include MAX_OVERFLOW and TOTAL.

And because these should be "private" (as private as python allows, anyway), they should start with an underscore - `_ASCII_TABLE`.

However I don't really love the idea of having a bunch of _SHOUTY_CAPS variables showing up in the module dictionary, especially given how many there are.

Maybe we could add all the tables to one big dictionary?

```python
_TABLE = {
    "ascii": Codepointrange({ascii}),
    ...
}
```

Note to self: We should also define `__all__` to only export wcwidth and Special by default (if you do `from widechar_width import *`).

Reply:
changes look good, I like the table